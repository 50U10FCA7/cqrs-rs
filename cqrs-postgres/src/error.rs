use cqrs_core::CqrsError;
use std::fmt;

/// An error while attempting to persist an event or snapshot.
#[derive(Debug)]
pub enum PersistError<E: CqrsError> {
    /// An error from the PostgreSQL backend.
    Postgres(postgres::Error),

    /// The operation failed because a specified precondition failed.
    PreconditionFailed(cqrs_core::Precondition),

    /// The operation failed because there was a serialization error.
    SerializationError(E),
}

impl<E: CqrsError> fmt::Display for PersistError<E> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            PersistError::Postgres(ref e) => write!(f, "postgres error: {}", e),
            PersistError::PreconditionFailed(ref e) => write!(f, "precondition error: {}", e),
            PersistError::SerializationError(ref e) => write!(f, "serialization error: {}", e),
        }
    }
}

impl<E: CqrsError> From<postgres::Error> for PersistError<E> {
    fn from(err: postgres::Error) -> Self {
        PersistError::Postgres(err)
    }
}

impl<E: CqrsError> From<cqrs_core::Precondition> for PersistError<E> {
    fn from(precondition: cqrs_core::Precondition) -> Self {
        PersistError::PreconditionFailed(precondition)
    }
}

/// An error while attempting to load an event or snapshot.
#[derive(Debug)]
pub enum LoadError<E: CqrsError> {
    /// An error from the PostgreSQL backend.
    Postgres(postgres::Error),

    /// The event type from the event stream is not one that can be deserialized.
    UnknownEventType(String),

    /// The operation failed because there was a deserialization error.
    DeserializationError(E),
}

impl<E: CqrsError> fmt::Display for LoadError<E> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            LoadError::Postgres(ref e) => write!(f, "postgres error: {}", e),
            LoadError::DeserializationError(ref e) => write!(f, "deserialization error: {}", e),
            LoadError::UnknownEventType(ref s) => write!(f, "unknown event type: {}", s),
        }
    }
}

impl<E: CqrsError> From<postgres::Error> for LoadError<E> {
    fn from(err: postgres::Error) -> Self {
        LoadError::Postgres(err)
    }
}
