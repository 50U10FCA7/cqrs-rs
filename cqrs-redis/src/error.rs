use std::fmt;

#[derive(Debug)]
pub enum PersistError {
    Redis(redis::RedisError),
    Serialization(rmp_serde::encode::Error),
    PreconditionFailed(cqrs_core::Precondition),
}

impl fmt::Display for PersistError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            PersistError::Redis(ref e) => write!(f, "redis error: {}", e),
            PersistError::Serialization(ref e) => write!(f, "serialization error: {}", e),
            PersistError::PreconditionFailed(ref e) => write!(f, "precondition error: {}", e),
        }
    }
}

impl From<redis::RedisError> for PersistError {
    fn from(err: redis::RedisError) -> Self {
        PersistError::Redis(err)
    }
}

impl From<rmp_serde::encode::Error> for PersistError {
    fn from(err: rmp_serde::encode::Error) -> Self {
        PersistError::Serialization(err)
    }
}

impl From<cqrs_core::Precondition> for PersistError {
    fn from(precondition: cqrs_core::Precondition) -> Self {
        PersistError::PreconditionFailed(precondition)
    }
}

#[derive(Debug)]
pub enum LoadError
{
    Redis(redis::RedisError),
    Deserialization(rmp_serde::decode::Error),
}

impl fmt::Display for LoadError
{
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            LoadError::Redis(ref e) => write!(f, "redis error: {}", e),
            LoadError::Deserialization(ref e) => write!(f, "deserialization error: {}", e),
        }
    }
}

impl From<redis::RedisError> for LoadError
{
    fn from(err: redis::RedisError) -> Self {
        LoadError::Redis(err)
    }
}

impl From<rmp_serde::decode::Error> for LoadError {
    fn from(err: rmp_serde::decode::Error) -> Self {
        LoadError::Deserialization(err)
    }
}
