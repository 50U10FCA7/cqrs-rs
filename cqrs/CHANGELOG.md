# master

* No changes yet

# [[0.1.0] 2019-02-01](https://github.com/cq-rs/cqrs/releases/tag/cqrs-0.1.0)

* Initial release
