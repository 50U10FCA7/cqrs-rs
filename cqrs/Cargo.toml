[package]
name = "cqrs"
version = "0.1.0-alpha.0"
authors = ["Marcus Griep <marcus@griep.us>"]
description = "An event-sourced command-query system"
license = "Apache-2.0"
edition = "2018"

[dependencies]
cqrs-core = { path = "../cqrs-core", version = "0.1.0-alpha.0" }
hashbrown = "0.1"
parking_lot = "0.7"
void = "1.0"

[dev-dependencies]
static_assertions = "0.3"
cqrs-todo-core = { path = "../cqrs-todo-core", version = "0.1.0-alpha.0" }

[badges]
travis-ci = { repository = "cq-rs/cqrs", branch = "master" }