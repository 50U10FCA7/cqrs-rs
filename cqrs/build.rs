extern crate rustc_version;

use rustc_version::Channel;

fn main() {
    let version = rustc_version::version_meta().unwrap();
    println!("cargo:rerun-if-changed-build=build.rs");
    if version.channel == Channel::Nightly {
        println!("cargo:rustc-cfg=nightly");
    }
}