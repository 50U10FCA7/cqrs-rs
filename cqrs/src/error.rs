use types::Precondition;
use std::error;
use std::fmt;

//#[derive(Debug, Clone, Hash, PartialEq)]
//pub enum LoadAggregateError<EErr, SErr> {
//    Snapshot(SErr),
//    Events(EErr),
//}
//
//impl<EErr, SErr> fmt::Display for LoadAggregateError<EErr, SErr>
//    where
//        EErr: error::Error,
//        SErr: error::Error,
//{
//    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
//        let err = self as &error::Error;
//        f.write_str(err.description())?;
//        f.write_str(": ")?;
//        write!(f, "{}", err.cause().unwrap())
//    }
//}
//
//impl<EErr, SErr> error::Error for LoadAggregateError<EErr, SErr>
//    where
//        EErr: error::Error,
//        SErr: error::Error,
//{
//    fn description(&self) -> &str {
//        match *self {
//            LoadAggregateError::Snapshot(_) => "loading snapshot",
//            LoadAggregateError::Events(_) => "loading events",
//        }
//    }
//
//    fn cause(&self) -> Option<&error::Error> {
//        match *self {
//            LoadAggregateError::Snapshot(ref e) => Some(e),
//            LoadAggregateError::Events(ref e) => Some(e),
//        }
//    }
//}
//
//#[derive(Debug, Clone, Hash, PartialEq)]
//pub enum PersistAggregateError<EErr, SErr> {
//    Snapshot(SErr),
//    Events(EErr),
//}
//
//impl<EErr, SErr> fmt::Display for PersistAggregateError<EErr, SErr>
//    where
//        EErr: error::Error,
//        SErr: error::Error,
//{
//    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
//        let err = self as &error::Error;
//        f.write_str(err.description())?;
//        f.write_str(": ")?;
//        write!(f, "{}", err.cause().unwrap())
//    }
//}
//
//impl<EErr, SErr> error::Error for PersistAggregateError<EErr, SErr>
//    where
//        EErr: error::Error,
//        SErr: error::Error,
//{
//    fn description(&self) -> &str {
//        match *self {
//            PersistAggregateError::Snapshot(_) => "storing snapshot",
//            PersistAggregateError::Events(_) => "storing events",
//        }
//    }
//
//    fn cause(&self) -> Option<&error::Error> {
//        match *self {
//            PersistAggregateError::Snapshot(ref e) => Some(e),
//            PersistAggregateError::Events(ref e) => Some(e),
//        }
//    }
//}
//
//#[derive(Debug, Clone, Hash, PartialEq)]
//pub enum ExecuteError<CErr, LErr> {
//    PreconditionFailed(AggregatePrecondition),
//    Command(CErr),
//    Load(LErr),
//}
//
//impl<CErr, LErr> fmt::Display for ExecuteError<CErr, LErr>
//    where
//        CErr: error::Error,
//        LErr: error::Error,
//{
//    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
//        if let ExecuteError::PreconditionFailed(ref precondition) = *self {
//            f.write_str("precondition failed: ")?;
//            match *precondition {
//                AggregatePrecondition::Exists => f.write_str("aggregate does not exist (expected existing)"),
//                AggregatePrecondition::New => f.write_str("aggregate already exists (expected new)"),
//                AggregatePrecondition::ExpectedVersion(AggregateVersion::Initial) => f.write_str("aggregate version does not match (expected default)"),
//                AggregatePrecondition::ExpectedVersion(AggregateVersion::Version(v)) => write!(f, "aggregate version does not match (expected version {})", v),
//            }
//        } else {
//            let err = self as &error::Error;
//            f.write_str(err.description())?;
//            if let Some(cause) = err.cause() {
//                write!(f, ": {}", cause)
//            } else {
//                Ok(())
//            }
//        }
//    }
//}
//
//impl<CErr, LErr> error::Error for ExecuteError<CErr, LErr>
//    where
//        CErr: error::Error,
//        LErr: error::Error,
//{
//    fn description(&self) -> &str {
//        match *self {
//            ExecuteError::PreconditionFailed(_) => "precondition failed",
//            ExecuteError::Command(_) => "invalid command",
//            ExecuteError::Load(_) => "loading aggregate",
//        }
//    }
//
//    fn cause(&self) -> Option<&error::Error> {
//        match *self {
//            ExecuteError::PreconditionFailed(_) => None,
//            ExecuteError::Command(ref e) => Some(e),
//            ExecuteError::Load(ref e) => Some(e),
//        }
//    }
//}
//
//impl<CErr, EErr, SErr> From<LoadAggregateError<EErr, SErr>> for ExecuteError<CErr, LoadAggregateError<EErr, SErr>>
//    where
//        CErr: error::Error,
//        EErr: error::Error,
//        SErr: error::Error,
//{
//    fn from(e: LoadAggregateError<EErr, SErr>) -> Self {
//        ExecuteError::Load(e)
//    }
//}
//
//#[derive(Debug, Clone, Hash, PartialEq)]
//pub enum ExecuteAndPersistError<EErr, PErr> {
//    Execute(EErr),
//    Persist(PErr),
//}
//
//impl<EErr, PErr> fmt::Display for ExecuteAndPersistError<EErr, PErr>
//    where
//        EErr: error::Error,
//        PErr: error::Error,
//{
//    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
//        let err = self as &error::Error;
//        f.write_str(err.description())?;
//        if let Some(cause) = err.cause() {
//            write!(f, ": {}", cause)
//        } else {
//            Ok(())
//        }
//    }
//}
//
//impl<EErr, PErr> error::Error for ExecuteAndPersistError<EErr, PErr>
//    where
//        EErr: error::Error,
//        PErr: error::Error,
//{
//    fn description(&self) -> &str {
//        match *self {
//            ExecuteAndPersistError::Execute(_) => "executing command",
//            ExecuteAndPersistError::Persist(_) => "persisting aggregate",
//        }
//    }
//
//    fn cause(&self) -> Option<&error::Error> {
//        match *self {
//            ExecuteAndPersistError::Execute(ref e) => Some(e),
//            ExecuteAndPersistError::Persist(ref e) => Some(e),
//        }
//    }
//}
//
//impl<XErr, EErr, SErr> From<PersistAggregateError<EErr, SErr>> for ExecuteAndPersistError<XErr, PersistAggregateError<EErr, SErr>>
//    where
//        XErr: error::Error,
//        EErr: error::Error,
//        SErr: error::Error,
//{
//    fn from(err: PersistAggregateError<EErr, SErr>) -> Self {
//        ExecuteAndPersistError::Persist(err)
//    }
//}
//
//impl<CErr, LErr, PErr> From<ExecuteError<CErr, LErr>> for ExecuteAndPersistError<ExecuteError<CErr, LErr>, PErr>
//    where
//        CErr: error::Error,
//        LErr: error::Error,
//        PErr: error::Error,
//{
//    fn from(err: ExecuteError<CErr, LErr>) -> Self {
//        ExecuteAndPersistError::Execute(err)
//    }
//}

#[derive(Debug, Clone, Hash, PartialEq)]
pub enum AppendEventsError<Err> {
    PreconditionFailed(Precondition),
    WriteError(Err),
}

impl<Err> fmt::Display for AppendEventsError<Err>
    where
        Err: error::Error,
{
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let err = self as &error::Error;
        f.write_str(err.description())?;
        f.write_str(": ")?;
        match *self {
            AppendEventsError::WriteError(ref e) => write!(f, "{}", e),
            AppendEventsError::PreconditionFailed(Precondition::ExpectedVersion(v)) => write!(f, "expected aggregate with version {}", v),
            AppendEventsError::PreconditionFailed(Precondition::New) => f.write_str("expected to create new aggregate"),
            AppendEventsError::PreconditionFailed(Precondition::Exists) => f.write_str("expected existing aggregate"),
        }
    }
}

impl<Err> error::Error for AppendEventsError<Err>
    where
        Err: error::Error,
{
    fn description(&self) -> &str {
        match *self {
            AppendEventsError::PreconditionFailed(_) => "precondition failed",
            AppendEventsError::WriteError(_) => "error appending events",
        }
    }

    fn cause(&self) -> Option<&error::Error> {
        match *self {
            AppendEventsError::WriteError(ref e) => Some(e),
            AppendEventsError::PreconditionFailed(_) => None,
        }
    }
}
impl<Err> From<Precondition> for AppendEventsError<Err> {
    fn from(p: Precondition) -> Self {
        AppendEventsError::PreconditionFailed(p)
    }
}