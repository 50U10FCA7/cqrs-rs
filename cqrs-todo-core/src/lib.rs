extern crate cqrs;
extern crate chrono;
extern crate smallvec;
extern crate serde;
#[macro_use] extern crate serde_derive;

use smallvec::SmallVec;
use cqrs::Aggregate;

pub mod domain {
    use chrono::{DateTime,Utc};
    use error::{InvalidDescription, InvalidReminderTime};
    use std::borrow::Borrow;

    #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
    pub struct Reminder {
        time: DateTime<Utc>,
    }

    impl Reminder {
        pub fn new(reminder_time: DateTime<Utc>, current_time: DateTime<Utc>) -> Result<Reminder, InvalidReminderTime> {
            if reminder_time <= current_time {
                Err(InvalidReminderTime)
            } else {
                Ok(Reminder {
                    time: reminder_time,
                })
            }
        }

        pub fn get_time(&self) -> DateTime<Utc> {
            self.time
        }
    }

    #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
    pub struct Description {
        text: String,
    }

    impl Description {
        pub fn new<S: Into<String>>(text: S) -> Result<Description, InvalidDescription> {
            let text = text.into();
            if text.is_empty() {
                Err(InvalidDescription)
            } else {
                Ok(Description {
                    text,
                })
            }
        }

        pub fn as_str(&self) -> &str {
            self.text.borrow()
        }
    }

    impl Borrow<str> for Description {
        fn borrow(&self) -> &str {
            self.text.borrow()
        }
    }
}

pub mod error {
    use std::error;
    use std::fmt;

    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
    pub struct InvalidReminderTime;

    impl error::Error for InvalidReminderTime {
        fn description(&self) -> &str {
            "reminder time cannot be in the past"
        }
    }

    impl fmt::Display for InvalidReminderTime {
        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
            let err = self as &error::Error;
            f.write_str(err.description())
        }
    }

    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
    pub struct InvalidDescription;

    impl error::Error for InvalidDescription {
        fn description(&self) -> &str {
            "description cannot be empty"
        }
    }

    impl fmt::Display for InvalidDescription {
        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
            let err = self as &error::Error;
            f.write_str(err.description())
        }
    }

    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
    pub enum CommandError {
        NotInitialized,
        AlreadyCreated,
    }

    impl error::Error for CommandError {
        fn description(&self) -> &str {
            match *self {
                CommandError::NotInitialized => "attempt to execute command before creation",
                CommandError::AlreadyCreated => "attempt to create when already created",
            }
        }
    }

    impl fmt::Display for CommandError {
        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
            let err = self as &error::Error;
            f.write_str(err.description())
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum Event {
    Created(domain::Description),
    TextUpdated(domain::Description),
    ReminderUpdated(Option<domain::Reminder>),
    Completed,
    Uncompleted,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Command {
    Create(domain::Description, Option<domain::Reminder>),
    UpdateText(domain::Description),
    SetReminder(domain::Reminder),
    CancelReminder,
    ToggleCompletion,
    MarkCompleted,
    ResetCompleted,
}

#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Serialize, Deserialize)]
pub enum TodoStatus {
    Completed,
    NotCompleted,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TodoData {
    pub description: domain::Description,
    pub reminder: Option<domain::Reminder>,
    pub status: TodoStatus,
}

impl TodoData {
    pub fn new(description: domain::Description, reminder: Option<domain::Reminder>, status: TodoStatus) -> Self {
        TodoData {
            description, reminder, status,
        }
    }

    pub fn with_description(description: domain::Description) -> Self {
        TodoData {
            description,
            reminder: None,
            status: TodoStatus::NotCompleted
        }
    }

    pub fn apply(&mut self, event: Event) {
        match event {
            Event::TextUpdated(description) => self.description = description,
            Event::ReminderUpdated(reminder) => self.reminder = reminder,
            Event::Completed => self.status = TodoStatus::Completed,
            Event::Uncompleted => self.status = TodoStatus::NotCompleted,
            Event::Created(_) => {},
        }
    }

    pub fn execute(&self, command: Command) -> Result<Events, error::CommandError> {
        match command {
            Command::UpdateText(description) => Ok(self.execute_update_description(description)),
            Command::SetReminder(reminder) => Ok(self.execute_set_reminder(reminder)),
            Command::CancelReminder => Ok(self.execute_cancel_reminder()),
            Command::ToggleCompletion => Ok(self.execute_toggle_completion()),
            Command::MarkCompleted => Ok(self.execute_mark_completed()),
            Command::ResetCompleted => Ok(self.execute_reset_completed()),
            Command::Create(description, reminder_opt) => self.execute_create(description, reminder_opt),
        }
    }

    pub fn execute_create(&self, description: domain::Description, reminder_opt: Option<domain::Reminder>) -> Result<Events, error::CommandError> {
        if self.description == description && self.reminder == reminder_opt {
            Ok(SmallVec::new())
        } else {
            Err(error::CommandError::AlreadyCreated)
        }
    }

    pub fn execute_update_description(&self, description: domain::Description) -> Events {
        let mut events = SmallVec::new();
        if self.description != description {
            events.push(Event::TextUpdated(description));
        }
        events
    }

    pub fn execute_set_reminder(&self, reminder: domain::Reminder) -> Events {
        let mut events = SmallVec::new();
        if let Some(existing) = self.reminder {
            if existing != reminder {
                events.push(Event::ReminderUpdated(Some(reminder)));
            }
        } else {
            events.push(Event::ReminderUpdated(Some(reminder)));
        }
        events
    }

    pub fn execute_cancel_reminder(&self) -> Events {
        let mut events = SmallVec::new();
        if self.reminder.is_some() {
            events.push(Event::ReminderUpdated(None));
        }
        events
    }

    pub fn execute_toggle_completion(&self) -> Events {
        let mut events = SmallVec::new();
        match self.status {
            TodoStatus::NotCompleted => events.push(Event::Completed),
            TodoStatus::Completed => events.push(Event::Uncompleted),
        }
        events
    }

    pub fn execute_mark_completed(&self) -> Events {
        let mut events = SmallVec::new();
        match self.status {
            TodoStatus::NotCompleted => events.push(Event::Completed),
            TodoStatus::Completed => {},
        }
        events
    }

    pub fn execute_reset_completed(&self) -> Events {
        let mut events = SmallVec::new();
        match self.status {
            TodoStatus::NotCompleted => {},
            TodoStatus::Completed => events.push(Event::Uncompleted),
        }
        events
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum TodoAggregate {
    Created(TodoData),
    Uninitialized,
}

impl Default for TodoAggregate {
    fn default() -> Self {
        TodoAggregate::Uninitialized
    }
}

type Events = SmallVec<[Event; 1]>;

impl TodoAggregate {
    pub fn apply_event(&mut self, event: Event) {
        match *self {
            TodoAggregate::Uninitialized => self.apply_to_uninitialized(event),
            TodoAggregate::Created(ref mut data) => data.apply(event),
        }
    }

    pub fn apply_to_uninitialized(&mut self, event: Event) {
        match event {
            Event::Created(initial_description) =>
                *self = TodoAggregate::Created(TodoData::with_description(initial_description)),
            _ => {}
        }
    }

    pub fn execute_command(&self, command: Command) -> Result<Events, error::CommandError> {
        match *self {
            TodoAggregate::Uninitialized => self.execute_on_uninitialized(command),
            TodoAggregate::Created(ref data) => data.execute(command),
        }
    }

    pub fn execute_on_uninitialized(&self, command: Command) -> Result<Events, error::CommandError> {
        match command {
            Command::Create(description, reminder_opt) => {
                let mut events = SmallVec::new();
                events.push(Event::Created(description));
                if let Some(reminder) = reminder_opt {
                    events.push(Event::ReminderUpdated(Some(reminder)));
                }
                Ok(events)
            }
            _ => Err(error::CommandError::NotInitialized),
        }
    }

    pub fn get_data(&self) -> Result<&TodoData, &'static str> {
        match *self {
            TodoAggregate::Uninitialized => Err("uninitialized"),
            TodoAggregate::Created(ref x) => Ok(x),
        }
    }
}

impl Aggregate for TodoAggregate {
    type Event = Event;
    type Command = Command;
    type Events = Events;
    type Error = error::CommandError;

    fn apply(&mut self, event: Self::Event) {
        println!("apply {:?}", event);
        self.apply_event(event);
    }

    fn execute(&self, command: Command) -> Result<Self::Events, Self::Error> {
        println!("execute {:?}", command);
        self.execute_command(command)
    }

    #[inline(always)]
    fn entity_type() -> &'static str where Self: Sized {
        "todo"
    }
}

#[cfg(test)]
mod tests {
    pub use super::*;
    use chrono::{Utc,TimeZone,Duration};

    fn create_basic_aggregate() -> TodoAggregate {
        let now = Utc.ymd(1970, 1, 1).and_hms(0, 0, 0);
        let reminder = now + Duration::seconds(10000);

        let mut events = SmallVec::<[Event;6]>::new();
        events.push(Event::Completed);
        events.push(Event::Created(domain::Description::new("Hello!").unwrap()));
        events.push(Event::ReminderUpdated(Some(domain::Reminder::new(reminder, now).unwrap())));
        events.push(Event::TextUpdated(domain::Description::new("New text").unwrap()));
        events.push(Event::Created(domain::Description::new("Ignored!").unwrap()));
        events.push(Event::ReminderUpdated(None));

        let mut agg = TodoAggregate::default();
        for event in events {
            agg.apply(event);
        }
        agg
    }

    #[test]
    fn example_event_sequence() {
        let expected_data = TodoData {
            description: domain::Description::new("New text").unwrap(),
            reminder: None,
            status: TodoStatus::NotCompleted,
        };
        let expected_state = TodoAggregate::Created(expected_data);

        let agg = create_basic_aggregate();

        assert_eq!(expected_state, agg);
    }

    #[test]
    fn cancel_reminder_on_default_aggregate() {
        let agg = TodoAggregate::default();

        let cmd = Command::CancelReminder;

        let result = agg.execute(cmd).unwrap_err();

        assert_eq!(error::CommandError::NotInitialized, result);
    }

    #[test]
    fn cancel_reminder_on_basic_aggregate() {
        let agg = create_basic_aggregate();

        let cmd = Command::CancelReminder;

        let result = agg.execute(cmd).unwrap();

        assert_eq!(Events::new(), result);
    }

    #[test]
    fn set_reminder_on_basic_aggregate() {
        let agg = create_basic_aggregate();

        let now = Utc.ymd(1970, 1, 1).and_hms(0, 0, 0);
        let reminder_time = now + Duration::seconds(20000);
        let reminder = domain::Reminder::new(reminder_time, now).unwrap();
        let cmd = Command::SetReminder(reminder);

        let result = agg.execute(cmd).unwrap();

        let mut expected = Events::new();
        expected.push(Event::ReminderUpdated(Some(reminder)));
        assert_eq!(expected, result);
    }
}
