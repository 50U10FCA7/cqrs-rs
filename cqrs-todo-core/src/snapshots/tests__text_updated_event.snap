---
created: "2019-01-24T17:38:40.675617+00:00"
creator: insta@0.5.2
expression: "RawEventWithName{event_name, raw: String::from_utf8(buffer).unwrap(),}"
source: cqrs-todo-core/src/lib.rs
---
RawEventWithName(
  event_name: "todo_text_updated",
  raw: "{\"text\":\"alt test description\"}",
)
